samples/qtask_unordered.c

```diff
@@ -0,0 +1,304 @@
+    }
+    if (!(in_samhdr = sam_hdr_read(infile))) {
+        printf("Failed to read header data\n");
+        goto end;
+    }
+    if (!(bamdata = bam_init1())) {
+        printf("Failed to initialize bamdata\n");
+        goto end;
+    }
+    size = strlen(data->region->name) + 50;         //region specification
+    if (!(region = malloc(size))) {
+        printf("Failed to allocate memory\n");
+        goto end;
+    }
+    snprintf(region, size, "%s:%"PRIhts_pos"-%"PRIhts_pos, data->region->name, data->region->start, data->region->end);
+    size += strlen(data->outdir);                   //output file with path
+    if (!(file = malloc(size))) {
+        printf("Failed to allocate memory\n");
+        goto end;
+    }
+    snprintf(file, size, "%s/%s_%"PRIhts_pos"_%"PRIhts_pos".sam", data->outdir, data->region->name, data->region->start, data->region->end);
+    if (!(idx = sam_index_load(infile, data->infile))) {
+        printf("Failed to load index\n");
+        goto end;
+    }
+    if (!(iter = sam_itr_querys(idx, in_samhdr, region))) {
+        printf("Failed to create iterator\n");
+        goto end;
+    }
+    if (!(outfile = sam_open(file, "w"))) {
+        printf("Failed to open output file\n");
+        goto end;
+    }
+    if (sam_hdr_write(outfile, in_samhdr) < 0) {
+        printf("Failed to write header\n");
+        goto end;
+    }
+    while ((ret = sam_itr_next(infile, iter, bamdata)) >= 0) {  //read and write relevant data
+        if (sam_write1(outfile, in_samhdr, bamdata) < 0) {
+            printf("Failed to write data\n");
+            goto end;
+        }
+    }
+    if (ret != -1) {
+        printf("Failed to get all data\n");
+    }
+
+end:
+    free(data);
+    if (infile) {
+        sam_close(infile);
+    }
+    if (outfile) {
+        sam_close(outfile);
+    }
+    if (iter) {
+        sam_itr_destroy(iter);
+    }
+    if (idx) {
+        hts_idx_destroy(idx);
+    }
+    if (bamdata) {
+        bam_destroy1(bamdata);
+    }
+    if (in_samhdr) {
+        sam_hdr_destroy(in_samhdr);
+    }
+    if (file) {
+        free(file);
+    }
+    if (region) {
+        free(region);
+    }
+    return NULL;
+}
+
+/// main - splits the data to region specific files
+/** @param argc - count of arguments
+ *  @param argv - pointer to array of arguments
+returns 1 on failure 0 on success
+*/
+int main(int argc, char *argv[])
+{
+    const char *inname = NULL, *outdir = NULL, *bedfile = NULL;
+    int c = 0, ret = EXIT_FAILURE, cnt = 0, regcnt = 0;
+    hts_tpool *pool = NULL;
+    hts_tpool_process *queue = NULL;
+    beddata *regions = NULL;
+
+    //qtask infile threadcount outdir [chunksize]
+    if (argc != 5) {
+        print_usage(stdout);
+        goto end;
+    }
+    inname = argv[1];
+    cnt = atoi(argv[2]);
+    outdir = argv[3];
+    bedfile = argv[4];
+    //get regions from bedfile
+    if ((regcnt = readbedfile(bedfile, &regions)) <= 0) {
+        printf("Failed to get bed data\n");
```

Comment:
I don't yet know why, but there's something non-deterministic about this:

```
$ for i in `seq 1 100`;do ./qtask_unordered ~/scratch/data/novaseq.10m.cram 8 /tmp/_jkb _.bed;done
Failed to get bed data
Failed to get bed data
$ more _.bed
1	1	3621
```

So out of 100 calls, it failed twice.  It's passing address sanitizer and thread sanitizer, but I'm suspecting another race condition somewhere.  It's rather odd.

Reply:
the separator for strtok_r was given as a const char instead of a const string and caused this failure. it is corrected.